receiver/mongodbreceiver/scraper.go

```diff
@@ -0,0 +1,171 @@
+	"go.mongodb.org/mongo-driver/bson"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver/internal/metadata"
+)
+
+const instrumentationLibraryName = "otelcol/mongodb"
+
+type mongodbScraper struct {
+	logger       *zap.Logger
+	config       *Config
+	client       client
+	mongoVersion *version.Version
+	mb           *metadata.MetricsBuilder
+}
+
+func newMongodbScraper(logger *zap.Logger, config *Config) *mongodbScraper {
+	return &mongodbScraper{
+		logger: logger,
+		config: config,
+		mb:     metadata.NewMetricsBuilder(config.Metrics),
+	}
+}
+
+func (s *mongodbScraper) start(ctx context.Context, _ component.Host) error {
+	clientLogger := s.logger.Named("mongo-scraper")
+	c := NewClient(s.config, clientLogger)
+	s.client = c
+
+	if err := s.client.Connect(ctx); err != nil {
+		s.logger.Error("unable to connect to mongo instance", zap.Error(err))
+	}
+
+	vr, err := s.client.GetVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to get a version from the mongo instance: %w", err)
+	}
+	v, err := version.NewVersion(*vr)
+	if err != nil {
+		return fmt.Errorf("unable to parse version as semantic version %s: %w", *vr, err)
+	}
+
+	s.mongoVersion = v
+
+	return nil
+}
+
+func (s *mongodbScraper) shutdown(ctx context.Context) error {
+	if s.client != nil {
+		return s.client.Disconnect(ctx)
+	}
+	return nil
+}
+
+func (s *mongodbScraper) scrape(ctx context.Context) (pdata.Metrics, error) {
+	s.logger.Debug("starting otelcol/mongodb scrape")
+	if s.client == nil {
+		return pdata.NewMetrics(), errors.New("no client was initialized before calling scrape")
+	}
+
+	metrics := pdata.NewMetrics()
+	rms := metrics.ResourceMetrics().AppendEmpty()
+	var errors scrapererror.ScrapeErrors
+
+	s.collectMetrics(ctx, rms, errors)
+
+	return metrics, errors.Combine()
+}
+
+func (s *mongodbScraper) collectMetrics(ctx context.Context, rms pdata.ResourceMetrics, errors scrapererror.ScrapeErrors) {
+	ilm := rms.InstrumentationLibraryMetrics().AppendEmpty()
+	ilm.InstrumentationLibrary().SetName(instrumentationLibraryName)
+
+	dbNames, err := s.client.ListDatabaseNames(ctx, bson.D{})
+	if err != nil {
+		s.logger.Error("Failed to fetch database names", zap.Error(err))
+		return
+	}
+
+	wg := &sync.WaitGroup{}
+
+	wg.Add(1)
+	go s.collectAdminDatabase(ctx, wg, errors)
+
+	for _, dbName := range dbNames {
+		wg.Add(1)
+		go s.collectDatabase(ctx, wg, dbName, errors)
+	}
+
+	wg.Wait()
+
+	s.mb.EmitCollection(ilm.Metrics())
+}
+
+func (s *mongodbScraper) collectDatabase(ctx context.Context, wg *sync.WaitGroup, databaseName string, errors scrapererror.ScrapeErrors) {
+	defer wg.Done()
+	dbStats, err := s.client.DBStats(ctx, databaseName)
+	if err != nil {
+		errors.AddPartial(1, err)
+	} else {
+		now := pdata.NewTimestampFromTime(time.Now())
```

Comment:
Typically we should try to have metrics align on a single timestamp, unless there's a good reason not to. You're getting time.now once for the admin db, and twice for each db, so the overall data set will have a lot of similar but slightly misaligned timestamps.

Reply:
Gotcha yeah the original idea was that the now's that were set were slightly more accurate based off the calls. I will go ahead and set the timestamp further up in the scrape in order to have a cohesive timestamp.